### ISSUE TITLE
Make sure we don't start compaction in the middle of loading

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#1368]
In issue #1366 @slivne noticed that we must not start compaction in the middle of loading a set of sstables belonging to the same CF (this can be either during boot, when we open all the sstables, or during a "nodetool refresh" operation doing load_new_sstables().

In commit faa45812b26dbed4e014d3f1cc4483de6da140a1 I fixed a case where this was happening in an "sstable rewrite" operation. But the question I want to raise here - could we also have the same bug with normal compactions started by the compaction manager?

I.e., can the compaction manager decide to start compaction for a CF before it is fully loaded? Can it start compaction in the middle of a load_new_sstables() operation?

**Comment by nyh** [comment]
Looking at the code, it appears that:
1. during load of the sstables at boot we don't have a problem, because compaction of a cf can only be triggered when a memtable is flushed for this cf, and this can only happen after the cf was fully loaded.
2. during nodetool refresh (load_new_sstables()), if in parallel new memtables are also being flushed, we can have compaction happen before being aware of all the added sstables. So I think this bug can exist. @raphaelsc what do you think?

Perhaps cf.run_with_compaction_disabled() method can be helpful here? But I think the compaction manager will also need to collect submit_sstable_rewrite() requests while disabled and run them when reenabled (this will replace the solution of commit faa4581, which kept the list of deferred compactions at the caller, not the compaction manager).

**Comment by raphaelsc** [comment]
@nyh: We can do the blocking steps earlier (load sstable but not add), and we only add new sstables to column family when done, atomically.

**Comment by avikivity** [comment]
Atomic add sounds good.

**Comment by avikivity** [comment]
Should do it for `nodetool refresh`, too.

**Comment by slivne** [comment]
@raphaelsc ping

**Comment by nyh** [comment]
@raphaelsc indeed sounds reasonable: column_family::load_new_sstables() could be split into two parts - first load all the sstables and collect them into a temporary array, and only then column_family::add_sstable() all of them.

There is a complication in that column_family::load_sstable() also has some logic of checking if the sstable is shared with other shards and therefore needs to be rewritten (split). This check will probably need to be reorganized.

**Comment by raphaelsc** [comment]
On Sun, Nov 6, 2016 at 9:10 AM, Shlomi Livne <notifications@github.com>
wrote:

> @raphaelsc <https://github.com/raphaelsc> ping
>
@slivne: patch sent
